### PR TITLE
Persist user context across arcs, rationalize Firebase utilization

### LINKED ISSUE #890: Deleted or updated shared entity data is not properly updated in BOXED_type.
Repro'd just now as:

- Have Walt open a new 'Write posts...' arc. Write two messages. Mark the arc as shared w/ friends.
- Have Scott open a new 'Show posts...' arc to view aggregated posts. You'll see the two messages from Walt.
- Have Walt delete one of the posts. Note that the deleted post still shows in Scott's arc. The logging seems to show that FireBase triggers an update, and the `BOXED_posts` are updated to include the one post still remaining from Walt, but that seems purely additive/update; that is, we never remove the deleted one from the existing `BOXED_posts` handle.

Log excerpt on Scott's arc side after Walt deletes the post below. The deleted post had message as 'Item One', and the remaining post had message as 'Item Two':

```
READING friend's [Walt] shared handles
remote-friends-shared-handles.js:151 RemoteFriendsSHs added friend's shared data to box [BOXED_blog_metadata] [{…}]
remote-friends-shared-handles.js:151 RemoteFriendsSHs added friend's shared data to box [BOXED_posts] 
[{…}]
0
:
id
:
"!987396311182881:demo-g51iznrav0rbb0uf:0:inner:5"
rawData
:
{author: "-L-8Ie9J417_6rBeMseD", createdTimestamp: 1520373548292, message: "Item Two.", owner: "-L-8Ie9J417_6rBeMseD"}
__proto__
:
Object
length
:
1
__proto__
:
Array(0)
````

And viewing the handle in Handle Explorer shows both the deleted item and the updated item above:

```
values
0
author
-L-8Ie9J417_6rBeMseD
createdTimestamp
1520288410816
message
An initial message.
owner
-L-8Ie9J417_6rBeMseD
1
author
-L-8Ie9J417_6rBeMseD
createdTimestamp
1520289667098
message
One more test.
owner
-L-8Ie9J417_6rBeMseD
2
author
-L-8Ie9J417_6rBeMseD
createdTimestamp
1520373542907
message
Item One.
owner
-L-8Ie9J417_6rBeMseD
3
author
-L-8Ie9J417_6rBeMseD
createdTimestamp
1520373548292
message
Item Two.
owner
-L-8Ie9J417_6rBeMseD
````

### LINKED ISSUE #1505: Launcher can get stuck flipping between two sets of Arcs tiles.
FYI @sjmiles perhaps not worth fixing until storage work is enabled but filing as FYI.

To repro:
- Open an incognito tab and visit https://polymerlabs.github.io/arcs-live/shell/apps/web/
- Wait until the launcher page loads and settles and tiles are visible
- Click the gear icon in bottom popup, click the profile icon, choose 'Walt'
- Don't click again (leave the popup open). Wait a bit. Eventually you should start seeing the tiles behind the partially opaque background start flipping between two sets. I saw this on both desktop and mobile.

### PR DESCRIPTION
**Pull request by sjmiles**
The _context_ is the collection of stores:

1. created by the shell (i.e. the list of user's arcs)
2. shared by the user
3. shared by the user's friends

This PR recognizes that the context belongs to the user, and only needs to be recomputed from scratch when the user changes.

Also, context data is dispersed through the Arcs Firebase and is highly dynamic. This PR introduces a new system for creating (live) queries against the database using a declarative structure ([FBGraph](https://drive.google.com/open?id=1I27XZXcp7ZgGKpJTP5kvQp8C4cvGQwAzXwGIgdj-2hA)).

The new query system is significantly faster and more robust than the existing ad-hoc code.

Should fix #1505, fix #890, and generally make everything work more smoothly.

### CODE REVIEW

File: shell/app-shell/elements/fb-data/FbStore.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import Arcs from '../../lib/arcs.js';
```
Comment (shaper) [inline shell/app-shell/elements/fb-data/FbStore.js]: Need copyright headers w/ 2018 here and a few other new files below.

File: shell/apps/utils/FBGraph/Field.js [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import {FbGraph} from '../../../components/FbGraph/FbGraph.js';
```
Comment (shaper) [inline shell/apps/utils/FBGraph/Field.js]: More copyright headers needed here and below.

File: shell/components/FbGraph/FbGraph.js [hunk 3]
```diff
@@ -0,0 +1,203 @@
+import Xen from '../xen/xen.js';
```
Comment (shaper) [inline shell/components/FbGraph/FbGraph.js]: Needs headers.

**Comment by sjmiles** [comment]
Last two commits were about fixing test issues.

**Pull request merged**